Sources/Plasma/FeatureLib/pfConsole/pfConsole.cpp

```diff
@@ -525,145 +527,76 @@ void    pfConsole::IHandleKey( plKeyEventMsg *msg )
-                    fPythonMultiLines = 0;
-                }
-            }
-            // else we are just doing single lines
-            else
-            {
-                // was there actually anything in the input buffer?
-                if ( fWorkingLine[0] != 0 )
-                {
-                    AddLineF(">>> %s", fWorkingLine);
-                    // check to see if this is going to be a multi line mode ( a ':' at the end)
-                    if ( fWorkingLine[eol-1] == ':' )
-                    {
-                        fPythonMultiLines = 1;
-                    }
-                    else
-                    // else if not the start of a multi-line then execute it
-                    {
-                        PyObject* mymod = PythonInterface::FindModule("__main__");
-                        PythonInterface::RunStringInteractive(fWorkingLine,mymod);
-                        std::string output;
-                        // get the messages
-                        PythonInterface::getOutputAndReset(&output);
-                        AddLine( output.c_str() );
-                    }
-                }
-                else
-                    AddLine( ">>> " );
-            }
-            // find the end of the line
-            for( c = fWorkingLine, eol = 0; *c != 0; eol++, c++ );
-        }
-        else
-        {
-            if( !fEngine->RunCommand( fWorkingLine, IAddLineCallback ) )
-            {
-                c = (char *)fEngine->GetErrorMsg();
-                if( c[ 0 ] != 0 )
-                    AddLine( c );
-            }
-        }
-
-        // Clear
-        fWorkingLine[ fWorkingCursor = 0 ] = 0;
+        IExecuteWorkingLine();
         findAgain = false;
         findCounter = 0;
         IUpdateTooltip();
     }
     else if( msg->GetKeyCode() == KEY_END )
     {
-        fWorkingCursor = strlen( fWorkingLine );
+        fWorkingCursor = strnlen(fWorkingLine, std::size(fWorkingLine));
     }
     else if( msg->GetKeyCode() == KEY_HOME )
     {
         fWorkingCursor = 0;
     }
+    else if (msg->GetCtrlKeyDown() && msg->GetKeyCode() == KEY_C)
+    {
+        // FIXME: console unicode friendly
+        plClipboard::GetInstance().SetClipboardText(ST::string::from_latin_1(fWorkingLine));
+    }
+    else if (msg->GetCtrlKeyDown() && msg->GetKeyCode() == KEY_X)
+    {
+        // FIXME: console unicode friendly
+        plClipboard::GetInstance().SetClipboardText(ST::string::from_latin_1(fWorkingLine));
+        fWorkingLine[fWorkingCursor = 0] = 0;
+        findAgain = false;
+        findCounter = 0;
+        IUpdateTooltip();
+    }
+    else if (msg->GetCtrlKeyDown() && msg->GetKeyCode() == KEY_V)
+    {
+        ST::string text = plClipboard::GetInstance().GetClipboardText();
+        if (text.empty())
+            return;
+
+        // Chop off leading or trailing newlines, which are probably not intended.
+        text = text.trim("\r\n");
+        if (text.empty())
+            return;
```

Comment:
Since `ST::string` instances are never null (right?), is it worth condensing these to do the `GetClipboardText()` and then trim before doing the if-empty-then-return bit? So then we're only checking once and not repeating?